### PR TITLE
Add NOREPLACE flag support for rename

### PR DESCRIPTION
**Pull request by co63oc**
https://github.com/asterinas/ccf-open-inno-competition

image
Add NOREPLACE flag support for rename.

Move the noreplace check to the interior of the write lock of DirDentry::rename, rather than the lock-free check at the syscall layer.
Eliminate TOCTOU race condition: The check is completed within the dentry write lock, and cannot be bypassed by other threads.
Covering cache miss: When the dentry cache misses, call inode.lookup() to query the actual file system.
sys_rename/sys_renameat: Pass false directly to maintain the original semantics.

### CODE REVIEW

File: kernel/src/fs/vfs/path/dentry.rs [hunk 1]
```diff
@@ -701,6 +709,15 @@ impl DirDentry<'_> {
 
             let mut children = old_dir.children.write();
             children.check_mountpoint(new_name)?;
+            if noreplace {
+                let target_exists = match children.find(new_name) {
+                    Some(entry) => entry.is_positive(),
+                    None => old_dir_inode.lookup(new_name).is_ok(),
+                };
+                if target_exists {
+                    return_errno_with_message!(Errno::EEXIST, "the new path already exists");
```
Comment (cqs21) [inline kernel/src/fs/vfs/path/dentry.rs]: In `DirDentry::rename`, the same-directory path has an early return:

```rust
if old_name == new_name {
    return Ok(());
}
```

This fires before the NOREPLACE check is reached. On Linux,
`renameat2(AT_FDCWD, "foo", AT_FDCWD, "foo", RENAME_NOREPLACE)` returns `EEXIST`.

File: kernel/src/fs/vfs/path/dentry.rs [hunk 2]
```diff
@@ -677,16 +677,24 @@ impl DirDentry<'_> {
...
         new_dir_arc: &Arc<Dentry>,
         new_name: &str,
+        noreplace: bool,
     ) -> Result<()> {
         let old_dir = old_dir_arc.as_dir_dentry_or_err()?;
         let new_dir = new_dir_arc.as_dir_dentry_or_err()?;
 
         if is_dot_or_dotdot(old_name) || is_dot_or_dotdot(new_name) {
+            if noreplace {
+                return_errno_with_message!(Errno::EINVAL, "old_name or new_name is a directory");
```
Comment (cqs21) [inline kernel/src/fs/vfs/path/dentry.rs]: The return value `EINVAL` is incorrect. When `new_name` is `.`/`..` and NOREPLACE is set, Linux returns `EEXIST`; for `old_name` it returns `EBUSY` regardless.

File: kernel/src/fs/vfs/path/dentry.rs [hunk 3]
```diff
@@ -677,16 +677,24 @@ impl DirDentry<'_> {
...
+    ///
+    /// When `noreplace` is true, the operation fails with `EEXIST` if the
+    /// target `new_name` already exists. This check is performed under the
+    /// dentry write lock to avoid TOCTOU races.
     pub(super) fn rename(
         old_dir_arc: &Arc<Dentry>,
         old_name: &str,
         new_dir_arc: &Arc<Dentry>,
         new_name: &str,
+        noreplace: bool,
```
Comment (cqs21) [inline kernel/src/fs/vfs/path/dentry.rs]: The coding guidelines state: "Avoid boolean arguments — use an enum or split into two functions." (`book/src/to-contribute/coding-guidelines/rust-guidelines/language-items/functions-and-methods.md#no-bool-args`) Uses a `RenameMode` enum instead:
```rust
pub enum RenameMode {
    Replace,
    NoReplace,
    Exchange,
}
```

This is more expressive, self-documenting, and directly extensible for future `EXCHANGE`/`WHITEOUT` support without a second refactor.

File: kernel/src/syscall/rename.rs [hunk 4]
```diff
@@ -28,8 +28,8 @@ pub fn sys_renameat2(
     let Some(flags) = Flags::from_bits(flags) else {
         return_errno_with_message!(Errno::EINVAL, "invalid flags");
     };
-    // TODO: Add support for handling the `NOREPLACE`, `EXCHANGE`, and `WHITEOUT` flags.
-    if !flags.is_empty() {
+    // TODO: Add support for handling the `EXCHANGE` and `WHITEOUT` flags.
+    if flags.intersects(Flags::EXCHANGE | Flags::WHITEOUT) {
```
Comment (cqs21) [inline kernel/src/syscall/rename.rs]: This conflates two distinct concerns: unsupported flags (`WHITEOUT`) and mutually exclusive flag combinations (`NOREPLACE | EXCHANGE`). Linux requires that `NOREPLACE` and `EXCHANGE` are mutually exclusive — combining them returns `EINVAL` regardless of whether `EXCHANGE` is implemented. Suggestion:
```rust
if flags.contains(Flags::WHITEOUT) || flags.contains(Flags::NOREPLACE | Flags::EXCHANGE) {
    return_errno_with_message!(Errno::EINVAL, "invalid renameat2 flag combination");
}
```

While the PR's `intersects` check happens to block the combined case (because `EXCHANGE` is in the mask), the intent is unclear and will cause confusion when `EXCHANGE` support is added later.

**Comment by cchanging** [comment]
It looks like this PR is included in #3148？

**Review by cqs21 — COMMENTED** [review]
Thanks for your contribution! I took a quick look and there are a few issues that
need to be addressed.

By the way, I previously added similar support (including `RENAME_EXCHANGE`) while
fixing LTP test cases in #3148 — feel free to use that as a reference. If possible,
it would be great to also include `RENAME_EXCHANGE` support in this PR.

**Comment by co63oc** [comment]
> Thanks for your contribution! I took a quick look and there are a few issues that need to be addressed.
> 
> By the way, I previously added similar support (including `RENAME_EXCHANGE`) while fixing LTP test cases in #3148 — feel free to use that as a reference. If possible, it would be great to also include `RENAME_EXCHANGE` support in this PR.


RENAME_EXCHANGE is a bit complicated, so for now I'll just add NOREPLACE. I will modify other options in other PR. The current CI has been completed. Please take some time to review it. Thank you. @cqs21

File: kernel/src/fs/vfs/path/dentry.rs (hunk 1 above)
Comment (co63oc) [reply kernel/src/fs/vfs/path/dentry.rs]: done

File: kernel/src/fs/vfs/path/dentry.rs (hunk 2 above)
Comment (co63oc) [reply kernel/src/fs/vfs/path/dentry.rs]: done

File: kernel/src/fs/vfs/path/dentry.rs (hunk 3 above)
Comment (co63oc) [reply kernel/src/fs/vfs/path/dentry.rs]: done

File: kernel/src/syscall/rename.rs (hunk 4 above)
Comment (co63oc) [reply kernel/src/syscall/rename.rs]: done

**Comment by cqs21** [comment]
> I will modify other options in other PR.

I'm working on 3148. Additionally, I would prefer enabling the LTP `renameat201` and `renameat202` tests directly rather than writing manual regression tests, unless the LTP tests fail to cover certain edge cases. However, enabling these two tests means you need to implement `RENAME_EXCHANGE` in this PR...

**Comment by co63oc** [comment]
> > I will modify other options in other PR.
> 
> I'm working on 3148. Additionally, I would prefer enabling the LTP `renameat201` and `renameat202` tests directly rather than writing manual regression tests, unless the LTP tests fail to cover certain edge cases. However, enabling these two tests means you need to implement `RENAME_EXCHANGE` in this PR...

Ok, that's a good idea.

**Comment by co63oc** [comment]
`renameat201` `renameat202` 已修改支持，增加 `RENAME_EXCHANGE` 支持

**Comment by tatetian** [comment]
@co63oc You submitted multiple PRs that span across unrelated kernel subsystems——epoll (#3395), fs (this PR), and network (#3404)——in last week. #3395 has been found to violate the best practices in Asterinas. This PR contains a non-trivial amount of changes to one of the core pieces of VFS, dentry. #3404 contains over 3K lines of changes. I don't think that a human fully understands this mount of code in multiple unrelated subsystems in such a short period of time. This practice violates our guidelines for participating the CCF open-source contribution competition ("AI is welcome, but human is responsible").

Similar to other open-source projects, we reserve the right to shut down or deny reviewing low-quality AI-generated PRs. For the reasons above, I am using this right to close this PR.

I will also close #3404. You are welcome to submit AI-assisted PRs in the future when you can fully understand and manage its complexity.

**Comment by co63oc** [comment]
> @co63oc You submitted multiple PRs that span across unrelated kernel subsystems——epoll (#3395), fs (this PR), and network (#3404)——in last week. #3395 has been found to violate the best practices in Asterinas. This PR contains a non-trivial amount of changes to one of the core pieces of VFS, dentry. #3404 contains over 3K lines of changes. I don't think that a human fully understands this mount of code in multiple unrelated subsystems in such a short period of time. This practice violates our guidelines for participating the CCF open-source contribution competition ("AI is welcome, but human is responsible").
> 
> Similar to other open-source projects, we reserve the right to shut down or deny reviewing low-quality AI-generated PRs. For the reasons above, I am using this right to close this PR.
> 
> I will also close #3404. You are welcome to submit AI-assisted PRs in the future when you can fully understand and manage its complexity.

Okay, thank you for your reply

**Comment by co63oc** [comment]
@cqs21 当前PR完整实现了 exfat, ext2的 `renameat201` `renameat202`支持，测试可以通过，https://github.com/asterinas/asterinas/pull/3148 中应该没有实现 exfat, ext2，可以参考下
